MAINT: Refactor special function ufunc generation and consolidate wrappers

izaid opened the pull request:
This PR merges _special.h into special_wrappers.cpp. There were a small number of functions in there. It also adds some extra testing for `sph_harm_all` for `m < -n`, which is just defined to be 0.

cc @steppi

Commit:
Extended tests for sph_harm to m < -n

Commit:
Merged _special.h into special_wrappers.cpp

Commit:
Removed exprel.pxd

steppi commented:
Thanks @izaid. It's definitely good to consolidate and not have so many wrappers files.

Commit:
Removed _logit.h

Commit:
.

Commit:
Added include <cmath>

Commit:
Templated exprel

Commit:
.

Commit:
.

steppi commented:
Failure on Linux Meson tests with gcc-8 is relevant.

Commit:
Made npy_type clearer and introduced npy_typenum

Commit:
.

Commit:
Bit of refactoring for ufunc.h

Commit:
.

Commit:
.

izaid commented:
Right, I hope this is done now, let's see on CI. It got a bit more involved as some of the log and exp related-functions involve `long double`, whose behaviour is a bit platform-dependent. I took that as an opportunity to harmonise a few other things.

Commit:
Cleaned up rank metaprogramming for gufuncs

izaid commented:
@steppi This is done.

Commit:
Last metaprogramming clean-up

Commit:
Created initializer_accumulate

Commit:
updated .clang-format

Commit:
.

h-vetinari commented:
> Failure on Linux Meson tests with gcc-8 is relevant.

We're just in the process of [updating](https://github.com/scipy/scipy/pull/20454) the compiler baseline requirements; we can easily move it to GCC 9, and if there's a concrete need, perhaps also to GCC 10. Since this is very recent (PR is not even merged yet), our CI hasn't been updated yet to reflect that, but JFYI in case this helps.

h-vetinari commented on scipy/special/special/.clang-format:
```diff
@@ -8,4 +8,5 @@ NamespaceIndentation: Inner
 ColumnLimit: 120
-InsertNewlineAtEOF: true
\ No newline at end of file
+InsertNewlineAtEOF: true
+AlignAfterOpenBracket: BlockIndent
```
The irony 😆

Commit:
.

Commit:
Updated ufunc initialization

Commit:
.

Commit:
.

Commit:
.

Commit:
switched to memcpy

Commit:
.

Commit:
Simplified SpecFun_Func template

steppi commented on scipy/special/special/legendre.h:
```diff
@@ -2,6 +2,8 @@
 
 #include "config.h"
 
+extern "C" double cephes_poch(double x, double m);
+
```
You know `poch` is already translated right? https://github.com/scipy/scipy/blob/main/scipy/special/special/cephes/poch.h

izaid replied:
I didn't know that! Great, done.

steppi commented on scipy/special/special/hyp2f1.h:
```diff
@@ -49,6 +49,8 @@
 #include "cephes/poch.h"
 #include "digamma.h"
 
+extern "C" double cephes_hyp2f1(double, double, double, double);
+
```
I don't like having this `extern "C"` from outside cephes thrown into what should be a self contained library. I think it's fine to just copy the `hyp2f1` from here into your PR.

https://github.com/scipy/scipy/blob/820ed56e237cd7091a2c4d4a5b5f89a7ef993975/scipy/special/special/cephes/hyp2f1.h

izaid replied:
Done.

steppi commented on scipy/special/special_wrappers.cpp:
```diff
@@ -208,9 +216,11 @@ void modified_fresnel_minus_wrap(double x, npy_cdouble *Fminus, npy_cdouble *Kmi
                                     reinterpret_cast<complex<double> *>(Kminus));
 }
 
-double sin_pi(double x) { return special::sin_pi(x); }
+double special_sinpi(double x) { return special::sinpi(x); }
+
+npy_cdouble special_csinpi(npy_cdouble z) { return to_ccomplex(special::sinpi(to_complex(z))); }
 
-double cos_pi(double x) { return special::cos_pi(x); }
+double special_cospi(double x) { return special::cos_pi(x); }
```
You seem to be using the `amos` `cospi` here instead of the `cephes` version which is currently the version SciPy uses for its `cospi` ufunc. Unless you've seen empirical evidence otherwise, I think the `cephes` version is superior.

izaid replied:
I think that's what it is doing before? In any case, switched to the cephes one.

steppi replied:
https://github.com/scipy/scipy/blob/438e88f5e1752f49b5c760e15c94c48f26d83e4d/scipy/special/functions.json#L12-L19

izaid replied:
Indeed, so it wasn't.

Though `special::cos_pi` did preexist, https://github.com/scipy/scipy/blob/main/scipy/special/special_wrappers.cpp#L213, no doubt from one of the many different wrappers / whatever.

Anyway, changed.

steppi commented on scipy/special/tests/test_sph_harm.py:
```diff
@@ -41,16 +41,21 @@ def Y11(theta, phi):
-                np.testing.assert_allclose(y_actual[-m, n], y_desired, rtol = 1e-05)
+            else:
+                y_desired = 0
+            np.testing.assert_allclose(y_actual[-m, n], y_desired, rtol = 1e-05)
```
`\n`

Commit:
Changes in response to review

steppi commented:
@izaid, I had added this to `config.h` in the cephes PR

```C++
#include <algorithm>
#include <cassert>
#include <cmath>
#include <complex>
#include <cstdint>
#include <limits>
#include <math.h>
#include <type_traits>

#ifdef DEBUG
#define SPECFUN_ASSERT(a) assert(a)
#else
#define SPECFUN_ASSERT(a)
#endif
```

and then instead of a direct `assert` in the code, it should be `SPECFUN_ASSERT`

Commit:
added cassert

Commit:
More fixes for assert

izaid commented:
> @izaid, for the failure for `hyp2f1.h`. I had added this to `config.h` in the cephes PR
> 
> ```c++
> #include <algorithm>
> #include <cassert>
> #include <cmath>
> #include <complex>
> #include <cstdint>
> #include <limits>
> #include <math.h>
> #include <type_traits>
> 
> #ifdef DEBUG
> #define SPECFUN_ASSERT(a) assert(a)
> #else
> #define SPECFUN_ASSERT(a)
> #endif
> ```
> 
> but instead of a direct `assert` in the code like I had, it should be `SPECFUN_ASSERT`.
> 
> The asserts were already there, so I figured I might as well offer a way to keep them.

Ok, put this in. The way you did `SPECFUN_ASSERT` is pretty normal for CUDA.

Commit:
changes for hyp2f1

steppi commented:
Sorry @izaid, I should have mentioned. You need to throw this do nothing `SPECFUN_ASSERT` into the CUDA section of the config too

https://github.com/scipy/scipy/blob/820ed56e237cd7091a2c4d4a5b5f89a7ef993975/scipy/special/special/config.h#L206

izaid commented:
> Sorry @izaid, I should have mentioned. You need to throw this do nothing `SPECFUN_ASSERT` into the CUDA section of the config too
> 
> https://github.com/scipy/scipy/blob/820ed56e237cd7091a2c4d4a5b5f89a7ef993975/scipy/special/special/config.h#L206

That's not in my PR, I think.

steppi commented:
> > Sorry @izaid, I should have mentioned. You need to throw this do nothing `SPECFUN_ASSERT` into the CUDA section of the config too
> > https://github.com/scipy/scipy/blob/820ed56e237cd7091a2c4d4a5b5f89a7ef993975/scipy/special/special/config.h#L206
> 
> That's not in my PR, I think.

Right, it's not, but I think it should be. It's from the entirety of cephes PR where I first defined `SPECFUN_ASSERT`. I asked you to copy that into this PR, but I think you should copy the CUDA part in too, so the symbol isn't undefined when running on CUDA.

Commit:
Added SPECFUN_ASSERT for CUDA

izaid commented:
> #define SPECFUN_ASSERT(a)

Ah, ok, I understood that you wanted me to delete that. Sure, I just put that in.

steppi commented:
@izaid, the changes to `ufunc.h` look pretty substantial, but there's no mention of them in the comments for this PR so far. Can you write out an overview here to give me an idea of what I should focus on?

izaid commented:
> @izaid, the changes to `ufunc.h` look pretty substantial, but there's no mention of them in the comments for this PR so far. Can you write out an overview here to give me an idea of what I should focus on?

Sure, there were a few things that happened. No changes to behaviour, more just tightening up of how things worked.

1. I completed the `npy_type` and `npy_typenum` structs so that they clearly map C++ types to NumPy type or typenums respectively. Before they weren't complete.

2. I refactored things a bit, introducing `npy_get` and `npy_set`. This already existed (basically), their purpose is to take the array data for a particular element and map it to whatever C++ needs. This could be as simple as casting, or it could be complex as putting it into an `std::mdspan` for the gufuncs. This code already existed, but I took care for a few edge cases (like to set `std::complex` properly).

3. `ufunc_data` here is now templated on the type of the function it stores. See https://github.com/scipy/scipy/blob/7599f31b91dd01c8aa115a7f58333af2ebedc1bc/scipy/special/ufunc.h#L350. Technically we're not supposed to be cast a function pointer to `void *` (though it often works), now we're not doing that at all.

4. I cleaned up some of the metaprogramming, in particular the offsets that needed to be added to `steps` in the `loop` function. Those used to be a lot more complicated.

5. I simplified the constructions of ufuncs at https://github.com/scipy/scipy/blob/7599f31b91dd01c8aa115a7f58333af2ebedc1bc/scipy/special/ufunc.h#L417, removing unncessary dynamic memory allocation and copying.

This was all kicked off because `npy_longdouble` is defined to be `npy_double` on some platforms, and `expit` is defined for `npy_longdouble`. That didn't quite work properly with what was there before, then I took the time to fix up some other things.

steppi commented:
Thanks, that's very helpful. I'll look it over a little more and will most likely merge tonight.

steppi reviewed (approved):
I'm happy with this @izaid. When you have a chance can you submit a PR to my `all_of_cephes` branch from here https://github.com/scipy/scipy/pull/20390 to clean up the resulting merge conflicts?
